src/org/thoughtcrime/securesms/FingerprintAuthenticationHandler.java

```diff
@@ -0,0 +1,213 @@
+
+    @Override
+    public void onAuthenticationFailed() {
+        Toast.makeText(context, R.string.prompt_passphrase_activity__fingerprint_auth_fail, Toast.LENGTH_LONG).show();
+    }
+
+    public void initializeFingerprintResources() {
+        try{
+            keyPairGenerator = getKeyPairGenerator();
+            cipher = getCipher();
+            keyStore = getKeyStore();
+
+            fingerprintManager = (FingerprintManager) context.getSystemService(FINGERPRINT_SERVICE);
+            if (initCipher(Cipher.DECRYPT_MODE)) {
+                cryptoObject = new FingerprintManager.CryptoObject(cipher);
+                beginAuthentication(fingerprintManager, cryptoObject);
+            }
+        } catch (FingerprintAuthenticationHandler.FingerprintException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void handlePassphraseChange(String passphrase){
+        try{
+            keyPairGenerator = getKeyPairGenerator();
+            cipher = getCipher();
+            keyStore = getKeyStore();
+            createKeyPair();
+            encrypt(passphrase);
+        } catch (FingerprintAuthenticationHandler.FingerprintException e) {
+            e.printStackTrace();
+        }
+    }
+    private void createKeyPair()  throws FingerprintAuthenticationHandler.FingerprintException {
+        try {
+            keyPairGenerator.initialize(
+                    new KeyGenParameterSpec.Builder(KEY_NAME, KeyProperties.PURPOSE_DECRYPT)
+                            .setDigests(KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA512)
+                            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_OAEP)
+                            .setUserAuthenticationRequired(true)
+                            .build());
+            keyPairGenerator.generateKeyPair();
+        } catch(InvalidAlgorithmParameterException exception) {
+            throw new FingerprintAuthenticationHandler.FingerprintException( exception);
+        }
+    }
+    public boolean initCipher(int opmode) throws FingerprintException {
+        try {
+            keyStore.load(null);
+            if(opmode == Cipher.ENCRYPT_MODE) {
+                PublicKey key = keyStore.getCertificate(KEY_NAME).getPublicKey();
+                PublicKey unrestricted = KeyFactory.getInstance(key.getAlgorithm()).generatePublic(new X509EncodedKeySpec(key.getEncoded()));
+                OAEPParameterSpec spec = new OAEPParameterSpec("SHA-256", "MGF1", MGF1ParameterSpec.SHA1, PSource.PSpecified.DEFAULT);
+                cipher.init(opmode, unrestricted, spec);
+            } else {
+                PrivateKey key = (PrivateKey) keyStore.getKey(KEY_NAME, null);
+                cipher.init(opmode, key);
+            }
+            return true;
+        } catch (NoSuchAlgorithmException | CertificateException | IOException | UnrecoverableKeyException | InvalidKeyException | InvalidKeySpecException | InvalidAlgorithmParameterException  | KeyStoreException e) {
```

Comment:
I think at this point it would look cleaner to catch either 1/a few common superclasses of these exceptions, if any, else just the general Exception

Reply:
Generally, I want the exceptions you catch to be as specific as possible so I don't suppress errors/bugs. I don't want to catch all Exceptions because it would suppress null pointers and other run time failures. I'll split it up on more lines to make it look cleaner.

Reply:
I agree, multiple lines would make this a lot more readable.

Reply:
I agree with keeping the different exception types. Splitting it onto multiple lines will make it more readable without losing information.